TAN-2224 Co-sponsors side effects

IvaKop opened the pull request:
# Changelog
<!-- Replace this comment by a bullet list. More info: https://www.notion.so/citizenlab/Changelog-How-it-works-f418426c75994454a332bf067634f3f1 -->

IvaKop committed:
Add cosponsorship activity

Commit:
Translations updated by CI (extract-intl)

IvaKop committed:
Add notification

IvaKop committed:
Merge branch 'TAN-2224/co-sponsors-side-effects' of github.com:CitizenLabDotCo/citizenlab into TAN-2224/co-sponsors-side-effects

IvaKop committed:
Add serializer

IvaKop committed:
Add notification front-end

Commit:
Translations updated by CI (extract-intl)

IvaKop committed:
Add accept notification

IvaKop committed:
Merge branch 'TAN-2224/co-sponsors-side-effects' of github.com:CitizenLabDotCo/citizenlab into TAN-2224/co-sponsors-side-effects

IvaKop committed:
Add notification front-end

Commit:
Translations updated by CI (extract-intl)

IvaKop committed:
Add factory

IvaKop committed:
Merge branch 'TAN-2224/co-sponsors-side-effects' of github.com:CitizenLabDotCo/citizenlab into TAN-2224/co-sponsors-side-effects

IvaKop committed:
Fix blint

IvaKop committed:
Rename file

IvaKop commented:
@kogre Re-creating the PR as the prev one got closed after I deleted the cosponsorship branch. There are still some relevant comment in there you might want to have a look at. To speed things up a bit main question is about the way the co-sposhorship ids diff is being handled and whether this can be done better.

kogre commented on back/app/services/side_fx_cosponsorship_service.rb:
```diff
@@ -0,0 +1,20 @@
+  def after_accept(cosponsorship, user)
+    LogActivityJob.perform_later(
+      cosponsorship,
+      'input_cosponsorship_accepted',
```
Detail, but why not go with `accepted` as the action name? The item and its type (Cosponsorship) is already part of the data.

kogre commented on back/app/controllers/web_api/v1/cosponsorships_controller.rb:
```diff
@@ -19,7 +19,7 @@ def index
   def accept_cosponsorship
     if @cosponsorship.status == 'pending'
       @cosponsorship.update(status: 'accepted')
```
Aware this is from the previous PR, but I missed it there. There is a check missing on whether the `update` operation succeeds. then, the else branch of _that_ check should call `render: json: { errors:....`. Currently, there will be no error for the status returned and if the `update` fails (that can be for any reason, maybe future code changes or the data becoming invalid in some way) there will be no error message returned.

An elegant way of solving this is to move the `status == pending` check to the model, but make it a conditional validation that is only called when doing this accept. By doing so, it ends up in your model errors and you only need to check whether the  `update` operations succeeds. See for example 
[this cnotroller](https://github.com/CitizenLabDotCo/citizenlab/blob/1c214c12f68cc4a53842fefdd3b40cc852bdd7a9/back/app/controllers/web_api/v1/invites_controller.rb#L138) in combination with [this model validation](https://github.com/CitizenLabDotCo/citizenlab/blob/1c214c12f68cc4a53842fefdd3b40cc852bdd7a9/back/app/models/user.rb#L183-L184)

IvaKop replied:
It was taking a bit of time to figure out this setup so I added the check directly into the controller as I've seen we do the vast majority of cases. Feel free to modify it in your PR, though, if you feel this is not good enough. :)

kogre commented on back/spec/models/notifications/cosponsor_of_your_idea_spec.rb:
```diff
@@ -0,0 +1,25 @@
+        action: 'input_cosponsorship_accepted'
+      )
+
+      notifications = described_class.make_notifications_on activity
```
Detail, but probably a good idea to also check that there is only a single notification in the returned array, just in case.

kogre commented on back/spec/models/notifications/invitation_to_cosponsor_idea_spec.rb:
```diff
@@ -0,0 +1,20 @@
+      cosponsorship = create(:cosponsorship, idea: idea, user: create(:user))
+      activity = create(:activity, item: cosponsorship, user_id: idea.author.id, action: 'created')
+
+      notifications = described_class.make_notifications_on activity
```
same

kogre commented on front/app/containers/MainHeader/Components/NotificationMenu/components/InvitationToCosponsorIdeaNotification/index.tsx:
```diff
@@ -0,0 +1,77 @@
+            question: messages.invitationToCosponsorQuestion,
+            issue: messages.invitationToCosponsorIssue,
+            contribution: messages.invitationToCosponsorContribution,
+            // initiative: messages.invitationToCosponsorInitiative,
+            // petition: messages.invitationToCosponsorPetition,
+            // proposal: messages.invitationToCosponsorProposal,
+          })}
```
Don't forget :)

kogre reviewed (approved):
Nice! Some small remarks, but nothing major.

You should also still update the ideas_spec acceptance tests. They should include the new cosponsors_ids in the params list for create and update, maybe also cover the case where we update the consponsors, as the controller plays an important role in that with those cosponsorship ids, so currently there's no coverage of the whole thing not breaking.

I think your solution for detecting the cosponsorships changes is sensible, and robust. The alternative would be to depend on [ActiveModel::Dirty](https://www.google.com/url?sa=t&source=web&rct=j&opi=89978449&url=https://api.rubyonrails.org/classes/ActiveModel/Dirty.html&ved=2ahUKEwjQqL3T1eyIAxVMVqQEHYA_CSMQFnoECBUQAQ&usg=AOvVaw3H1lYshz4X3U8aoZX1XF5l), which since Rails 7 [also supports associations](https://blog.saeloun.com/2021/07/27/rails-7-belongs-to-change-tracking/). When the original initiatives code was written, we weren't on Rails 7 yet, I believe. The advantage of that approach is that it's fully controlled in de side fx, instead of leaking into the controller. That being said, I think this is pretty simple and more robust, so I'm okay with it, don't think it's worth changing.

IvaKop committed:
Replace action

IvaKop committed:
Check if only one notification

IvaKop committed:
Add missing spec

IvaKop committed:
Merge branch 'master' into TAN-2224/co-sponsors-side-effects

IvaKop committed:
Add to acceptance spec

IvaKop committed:
Merge branch 'TAN-2224/co-sponsors-side-effects' of github.com:CitizenLabDotCo/citizenlab into TAN-2224/co-sponsors-side-effects

IvaKop commented:
@kogre I've addressed your comments, thanks for reviewing! I still couldn't figure out how to fix the failing specs.. 🤔 
Would be great if you could take a look and let me know what I'm missing 🙌

kogre committed:
Added cosponsorship email campaigns

kogre committed:
Solve logging of idea assignee changed activity

kogre committed:
Merge pull request #9040 from CitizenLabDotCo/TAN-2224/co-sponsors-emails

TAN-2224/co sponsors emails

kogre commented:
@IvaKop The reason why the test fails is because the default state of the cosponsor_ids custom_field is set to `enabled: false`, which means that eventually, it isn't part of the allowed parameters in the controller.

https://github.com/CitizenLabDotCo/citizenlab/blob/00ad8950513946856a95731c3c52ec52472dd0ce/back/lib/participation_method/ideation.rb#L325

I'm not fully sure what the implcation of the `enabled` field on the form is wrt to the form builder. Either we should change this boolean, or the failing spec should manually configure the custom field to be enabled.

IvaKop commented:
@kogre I find this confusing because this is overridden and set to true when creating a proposal so I'm not sure how to solve it. Let's maybe take a look together.

kogre committed:
Fixed idea create spec

kogre commented:
@IvaKop It's fixed, we had a factory in place to easily create a custom form with the default fields, so I was able to use that, in the end.

IvaKop commented:
Thanks @kogre! Clean solution in the end 👍 Really appreciate your help on this one 🙌

IvaKop committed:
Merge branch 'master' into TAN-2224/co-sponsors-side-effects
